element-hq/hydrogen-web
# Add license comments to make GNU LibreJS accept Hydrogen's JS

This adds proper license comments to all _internal_ JS scripts in the `/target` build. This way the code self-identifies as Apache-2.0 which prevents LibreJS from blocking it.

Given the [supported options](https://www.gnu.org/software/librejs/free-your-javascript.html), dynamically adding `// @license ...` / `// @license-end` comments via `build.mjs` seemed like the most practical solution. This only makes the `/target` build work with LibreJS not the development build but achieving the latter would have required annotation _all_ JS source files.

Note that this fix doesn't do anything about _external_ scripts, namely olm. It would be possible to declare olm's license on the Hydrogen level, too. However, I believe this should rather be addressed upstream so that any client that uses olm can benefit from it being LibreJS-compliant (see matrix-org/olm#52).

To test:

- Install the [LibreJS add-on for Firefox](https://addons.mozilla.org/en-US/firefox/addon/librejs/)
- `yarn build && yarn start`
- Load `localhost:3000/target` and notice how all scripts except for olm have been allowed by LibreJS

![Screenshot 2021-03-27 at 14 28 13](https://user-images.githubusercontent.com/1137962/112722256-ad9a1a00-8f08-11eb-8763-280f7e2128d5.png)

To test the whole app, wrap the content of `target/olm-....js` with

```
// @license magnet:?xt=urn:btih:8e4f440f4c65981c5bf93c76d35135ba5064d8b7&dn=apache-2.0.txt Apache-2.0
// @source: https://gitlab.matrix.org/matrix-org/olm/-/tree/3.2.2
```

and

```
// @license-end
```

after building.

Closes: #209

Signed-off-by: Johannes Marbach <n0-0ne+github@mailbox.org>
